### PR TITLE
Consistent ordering of planning areas in PDF

### PR DESCRIPTION
**Pull request by andrea-williams**
# Test Links:
[Percentile Calculator](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=1108&f=c5&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN#state=2ec784ca-c46a-49d0-b2b3-1cf32a9015a2&session_state=7d9447c8-db66-4661-b4cb-03d2ac0d1d8f&code=32292df4-2bdf-4f90-a4a8-c8dbcda682a9.7d9447c8-db66-4661-b4cb-03d2ac0d1d8f.2b63f390-f3dc-43ae-89f2-016453863476)
[Fire Behaviour Advisory](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-1950.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

File: api/app/tests/hfi/test_pdf_formatter.py [hunk 1]
```diff
@@ -238,10 +238,5 @@ def test_response_2_daily_jinja_format():
...
             assert len(daily_planning_area_data) == 7
             # assert correct order for each day
-            assert daily_planning_area_data[0].planning_area_name == 'Fraser Zone'
-            assert daily_planning_area_data[1].planning_area_name == 'Pemberton Zone'
-            assert daily_planning_area_data[2].planning_area_name == 'Sunshine Coast'
-            assert daily_planning_area_data[3].planning_area_name == 'South Island'
-            assert daily_planning_area_data[4].planning_area_name == 'Mid Island'
-            assert daily_planning_area_data[5].planning_area_name == 'North Island'
-            assert daily_planning_area_data[6].planning_area_name == 'Mid-Coast'
+            for index in range(len(daily_planning_area_data)):
```
Comment (conbrad) [inline api/app/tests/hfi/test_pdf_formatter.py:241]: Nice!

**Review by conbrad — CHANGES_REQUESTED** [review]

**Review by Sybrand — CHANGES_REQUESTED** [review]
Nit picking variable naming.

**Comment by conbrad** [comment]
[KamloopsFireCentre_HFICalculator_2022-04-27_CBRADY@IDIR.pdf](https://github.com/bcgov/wps/files/8575111/KamloopsFireCentre_HFICalculator_2022-04-27_CBRADY%40IDIR.pdf)

^ Out of scope for this ticket, but it looks like we have the same problem with station ordering (e.g. Station Bay 2 is last in Vernon, but in HFI it's 3rd)

File: api/app/hfi/pdf_data_formatter.py [hunk 2]
```diff
@@ -174,15 +174,20 @@ def response_2_daily_jinja_format(result: HFIResultResponse,
...
                                                        fire_starts=daily_result.fire_starts.label,
                                                        date=daily_result.date.isoformat(),
                                                        dailies=station_daily_pdf_data)
             daily_pdf_data.append(daily_data)
 
-    key = operator.attrgetter('date')
+    order_by_date_key = operator.attrgetter('date')
+    daily_pdf_data.sort(key=lambda k: (k.date, k.order))
     daily_pdf_data_by_date = dict((k, list(map(lambda x: x, values)))
-                                  for k, values in groupby(sorted(daily_pdf_data, key=key), key))
```
[the remark this answers is not in the document]
Comment (andrea-williams) [reply api/app/hfi/pdf_data_formatter.py:187]: L188 - `daily_pdf_data.sort(key=lambda k: (k.date, k.order))` means that `daily_pdf_data` is first being sorted by date, and then within each date, it's being sorted by the `order` attribute.

[the remark this answers is not in the document]
Comment (conbrad) [reply api/app/hfi/pdf_data_formatter.py:187]: Hmm, but the test breaks in with the conditions I state above, please check.

**Review by conbrad — APPROVED** [review]

**Review by Sybrand — APPROVED** [review]
LGTM. But will mention - ideally we should have a unit test that catches this situation.

**Comment by andrea-williams** [comment]
> LGTM. But will mention - ideally we should have a unit test that catches this situation.

Do you mean that catches the bug this fixes? I modified the existing unit test so that we're comparing against the `order` value instead of just planning area names. Is there some other unit test I could set up?

**Comment by Sybrand** [comment]
> > LGTM. But will mention - ideally we should have a unit test that catches this situation.
> 
> Do you mean that catches the bug this fixes? I modified the existing unit test so that we're comparing against the `order` value instead of just planning area names. Is there some other unit test I could set up?

There's a unit test that checks the output from the HTML. The fact that it didn't fail, means that the data we're feeding the test wasn't exposing the ordering issue. Specifically, api/app/tests/hfi/test_fire_centres.json probably has the records in the json in an order that matches the order of appearance.

**Pull request merged**